File: app/src/main/java/me/imzack/app/cold/domain/activity/SettingsActivity.java
package me.imzack.app.cold.domain.activity;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.preference.SwitchPreference;
import android.support.v7.app.AppCompatDelegate;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import me.imzack.app.cold.R;
import me.imzack.app.cold.model.preference.PreferenceDispatcher;
import me.imzack.app.cold.interactor.presenter.SettingsPresenter;
import me.imzack.app.cold.util.LogUtil;
import me.imzack.app.cold.domain.view.SettingsView;

import java.util.Locale;

public class SettingsActivity extends BaseActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);

        setSupportActionBar(toolbar);
        setupActionBar();

        getFragmentManager().beginTransaction().replace(R.id.frame_layout, new SettingsFragment()).commit();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    public static class SettingsFragment extends PreferenceFragment
            implements SettingsView, SharedPreferences.OnSharedPreferenceChangeListener {

        private SettingsPresenter settingsPresenter;
        private SwitchPreference locationServicePref;
        private ListPreference languagePref, nightModePref;

        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.preferences);

            settingsPresenter = new SettingsPresenter(this);

            settingsPresenter.setInitialView();
        }

        @Override
        public void onResume() {
            super.onResume();
            getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        }

        @Override
        public void onPause() {
            super.onPause();
            getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        }

        @Override
        public void showInitialView() {
            locationServicePref = (SwitchPreference) findPreference(PreferenceDispatcher.KEY_PREF_LOCATION_SERVICE);
            languagePref = (ListPreference) findPreference(PreferenceDispatcher.KEY_PREF_LANGUAGE);
            nightModePref = (ListPreference) findPreference(PreferenceDispatcher.KEY_PREF_NIGHT_MODE);

            initPreferenceSummary();
        }

        @Override
        public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
            switch (key) {
                case PreferenceDispatcher.KEY_PREF_LOCATION_SERVICE:
                    if (sharedPreferences.getBoolean(PreferenceDispatcher.KEY_PREF_LOCATION_SERVICE, false)) {
                        //已打开定位服务
                        LogUtil.d("已打开定位服务");
                    } else {
                        //已关闭定位服务
                        LogUtil.d("已关闭定位服务");
                    }
                    break;
                case PreferenceDispatcher.KEY_PREF_LANGUAGE:
                    String languageValue = sharedPreferences.getString(PreferenceDispatcher.KEY_PREF_LANGUAGE, "");
                    languagePref.setSummary(languagePref.getEntries()[languagePref.findIndexOfValue(languageValue)]);

                    Configuration config = getResources().getConfiguration();
                    switch (languageValue) {
                        case "en":
                            config.locale = Locale.ENGLISH;
                            break;
                        case "zh":
                            config.locale = Locale.CHINESE;
                            break;
                        default:
                            config.locale = Locale.getDefault();
                            break;
                    }
                    getResources().updateConfiguration(config, null);
                    recreate();
                    break;
                case PreferenceDispatcher.KEY_PREF_NIGHT_MODE:
                    String nightModeValue = sharedPreferences.getString(PreferenceDispatcher.KEY_PREF_NIGHT_MODE, "");
                    nightModePref.setSummary(nightModePref.getEntries()[nightModePref.findIndexOfValue(nightModeValue)]);

                    int mode = AppCompatDelegate.MODE_NIGHT_NO;
                    switch (nightModeValue) {
                        case "off":
                            mode = AppCompatDelegate.MODE_NIGHT_NO;
                            break;
                        case "on":
                            mode = AppCompatDelegate.MODE_NIGHT_YES;
                            break;
                        case "auto":
                            mode = AppCompatDelegate.MODE_NIGHT_AUTO;
                            break;
                        case "def":
                            mode = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;
                            break;
                        default:
                            break;
                    }
                    AppCompatDelegate.setDefaultNightMode(mode);
                    recreate();
                    break;
                default:
                    break;
            }
        }

        private void initPreferenceSummary() {
            SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
            //Location Service
            //No need to set summary
            //Language
            languagePref.setSummary(languagePref.getEntries()[languagePref.findIndexOfValue(sharedPreferences.getString(PreferenceDispatcher.KEY_PREF_LANGUAGE, ""))]);
            //Night Mode
            nightModePref.setSummary(nightModePref.getEntries()[nightModePref.findIndexOfValue(sharedPreferences.getString(PreferenceDispatcher.KEY_PREF_NIGHT_MODE, ""))]);
        }

        /** 返回并重新创建HomeActivity */
        private void recreate() {
            Intent intent = new Intent(getActivity(), HomeActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
        }
    }
}

File: app/src/main/java/net/zackzhang/app/cold/view/activity/SettingsActivity.kt
package net.zackzhang.app.cold.view.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.Menu
import android.view.MenuItem
import kotlinx.android.synthetic.main.activity_settings.*
import net.zackzhang.app.cold.R
import net.zackzhang.app.cold.model.DataManager
import net.zackzhang.app.cold.view.dialog.MessageDialogFragment
import net.zackzhang.app.cold.view.fragment.SettingsFragment

class SettingsActivity : BaseActivity() {

    companion object {

        private const val TAG_RESET_SETTINGS = "reset_settings"

        fun start(context: Context) {
            context.startActivity(Intent(context, SettingsActivity::class.java))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        setSupportActionBar(vToolbar)
        setupActionBar()

        if (savedInstanceState == null) {
            // 说明 Activity 是新启动的，添加 SettingsFragment；否则 SettingsFragment 会自动恢复，不需要重复添加
            supportFragmentManager.beginTransaction().add(R.id.vContentLayout, SettingsFragment()).commit()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_settings, menu)
        return true
    }

    override fun onAttachFragment(fragment: Fragment) {
        super.onAttachFragment(fragment)

        if (fragment.tag == TAG_RESET_SETTINGS) {
            (fragment as MessageDialogFragment).okButtonClickListener = { DataManager.preferenceHelper.resetAllValues() }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> exit()
            R.id.action_reset -> MessageDialogFragment.Builder()
                    .setTitle(R.string.title_dialog_reset_settings)
                    .setMessage(R.string.msg_dialog_reset_settings)
                    .setOkButtonText(R.string.btn_dialog_reset_settings)
                    .showCancelButton()
                    .show(supportFragmentManager, TAG_RESET_SETTINGS)
        }
        return super.onOptionsItemSelected(item)
    }
}
